src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -73099,6 +73099,232 @@ static void Main()
+            var i2 = c2.GetTypeMember("I");
+            Assert.Equal("C<T>.I<U>", i2.ToTestDisplayString());
+            AssertTypesDifferButHashCodesMatch(i, i2);
+
+            var i2b = new ConstructedNamedTypeSymbol(i2, ImmutableArray.Create(TypeWithAnnotations.Create(i2.TypeParameters.Single(), NullableAnnotation.Annotated)));
+            Assert.Equal("C<T>.I<U?>", i2b.ToTestDisplayString());
+            AssertTypesDifferButHashCodesMatch(i, i2b);
+
+            var c3 = new ConstructedNamedTypeSymbol(c, ImmutableArray.Create(TypeWithAnnotations.Create(c.TypeParameters.Single(), NullableAnnotation.Annotated)));
+            var i3 = c3.GetTypeMember("I");
+            Assert.Equal("C<T?>.I<U>", i3.ToTestDisplayString());
+            AssertTypesDifferButHashCodesMatch(i, i3);
+
+            var i3b = new ConstructedNamedTypeSymbol(i3, ImmutableArray.Create(TypeWithAnnotations.Create(i3.TypeParameters.Single(), NullableAnnotation.Annotated)));
+            Assert.Equal("C<T?>.I<U?>", i3b.ToTestDisplayString());
+            AssertTypesDifferButHashCodesMatch(i, i3b);
+        }
+
+        private static void AssertTypesDifferButHashCodesMatch(TypeSymbol c, TypeSymbol c2)
+        {
+            Assert.False(c.Equals(c2));
+            Assert.True(c.Equals(c2, TypeCompareKind.CLRSignatureCompareOptions));
+
+            Assert.Equal(c2.GetHashCode(), c.GetHashCode());
+        }
+
+        [Fact]
+        [WorkItem(30677, "https://github.com/dotnet/roslyn/issues/30677")]
+        public void ExplicitInterface_UsingOuterDefinition_Simple()
+        {
+            var text = @"
+class Outer<T>
+{
+    protected internal interface Interface
+    {
+        void Method();
+    }
+    internal class C : Outer<T>.Interface
+    {
+        void Interface.Method()
+        {
+        }
+    }
+}
+";
+            // https://github.com/dotnet/roslyn/issues/30677 - Expect no warning
+            var comp = CreateNullableCompilation(text);
+            comp.VerifyDiagnostics(
+                // (10,14): warning CS8643: Nullability of reference types in explicit interface specifier doesn't match interface implemented by the type.
+                //         void Interface.Method()
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInExplicitlyImplementedInterface, "Interface").WithLocation(10, 14)
+                );
+        }
+
+        [Fact]
+        [WorkItem(30677, "https://github.com/dotnet/roslyn/issues/30677")]
+        public void ExplicitInterface_UsingOuterDefinition()
+        {
+            var text = @"
+class Outer<T> where T : class
+{
+    internal class Inner<U> where U : class
+    {
+        protected internal interface Interface
+        {
+            void Method();
+        }
+        // The implemented interface is Outer<T!>.Inner<U!>.Interface
+        internal class Derived6 : Outer<T>.Inner<U>.Interface
+        {
+            // The explicit interface is Outer<T>.Inner<U!>.Interface
```

Comment:
Tagging @AlekseyTs for discussion. 
This explicit interface uses the definition of `Outer<T>`, but I'm thinking it should use a fully constructed type instead: `Outer<T!>.Inner<U!>.Interface` as when the type is completely spelled out (in the implemented interface two lines above, or in the test `ExplicitInterface_WithExplicitOuter` below). 
What do you think?

Reply:
> This explicit interface uses the definition of Outer<T>, but I'm thinking it should use a fully constructed type instead: Outer<T!>.Inner<U!>.Interface as when the type is completely spelled out (in the implemented interface two lines above, or in the test ExplicitInterface_WithExplicitOuter below). 

I think this is something worth discussing. But I am not sure why ```Outer<T!>.Inner<U!>``` is a better choice and whether current annotation context should play a role the same way as it does when the name is spelled out. 


---
In reply to: [282670985](https://github.com/dotnet/roslyn/pull/35613#discussion_r282670985) [](ancestors = 282670985)

Reply:
Filed [https://github.com/dotnet/roslyn/issues/35619](https://github.com/dotnet/roslyn/issues/35619)

---
In reply to: [283579930](https://github.com/dotnet/roslyn/pull/35613#discussion_r283579930) [](ancestors = 283579930,282670985)